### ISSUE TITLE
Categories widget - HTML class "current-cat" attached to wrong category when on a single product page

### ISSUE THREAD

**Issue by cartible** [woothemes/woocommerce#10555]
## EXPLANATION OF THE ISSUE

The Product Categories widget uses the HTML class "current-cat" to indicate the current category that a product, group of products or child categories belong to.

The problem is when a single-product page is displayed, the top-most category instead of the immediate parent category of the product is attached with the "current-cat" HTML class.

This happens for the default WooCommerce version with no other plugins activated.

## STEPS TO REPRODUCE THE ISSUE

Assuming the use of the sample WooCommerce data, latest stable WooCommerce version and Store Front or default WordPress themes.

1. Click on the Clothing > Hoodies category in the Categories widget will bring up the category archive.

2. Choosing "Woo Ninja" will show the single-product page. 

3. Notice that the active category is "Clothing" and not "Hoodies" because the "current-cat" HTML class is attached to the "Clothing" category instead of the "Hoodies" category.   

## IDENTIFIED PROBLEM

The problem is due to WooCommerce selecting the top-most category on a single-product page as found in the following code in https://github.com/woothemes/woocommerce/blob/master/includes/widgets/class-wc-widget-product-categories.php#L127

```
$this->current_cat   = end( $product_category );
```

Changing the code to the following fixes the issue.

```
$this->current_cat   = reset( $product_category );
```

## CONCLUSION

I'm not sure whether this is a WooCommerce bug or feature but it makes more sense to have the immediate parent category highlighted instead of the top-most category highlighted as what one of my users asked.

**Comment by mikejolley** [comment]
We feel top most makes the most sense in the majority of cases, as its the 'primary' category for the product. If you have categories in your permalink structure, the same will be used there.

By using children you run into issues with complex category hierarchies e.g. if you assign a product to 2 child sibling categories.

This also affects;
- showing children of current cat (top most shows more relevant cats)
- permalinks
- breadcrumbs

Looks like this is the type of issue which can be desired to work either way, so changing this will upset part of the user base. I'd suggest if you want to control this we either introduce a filter on current cat, or on the woocommerce_get_product_terms call so you can change the order they are returned.

**Comment by cartible** [comment]
Thank you for shedding more light on the rationale behind this feature. 

I'd opt for a change in the `wc_get_product_terms()` to be made.

Much appreciated.

**Comment by pickmeok** [comment]
Hi @ghost and @mikejolley

I am facing the same issue with Woocommerce 4.3.1.

I am using the Woocommerce Products Category widget on the single product page.

I have set the a child category as the primary one. To the product in question I have assigned to it 3 child categories ans I expected the Woocommerce Products Category widget to display the primary categorya not the others. Breadcrubms and url at product page are correct.

I checked the code and the points you made but I could not fix it.

Any solution for this issue?
Thank you
